packages/core/src/connect.tsx

```diff
@@ -75,203 +87,148 @@ export type Func<V, P = {}> = (viewCtx: V) => React.ComponentType<P>;
+    );
 
-      Object.keys(this.prevWatched).forEach(pathKey => {
-        const keyDeleted = !this.watched.hasOwnProperty(pathKey);
-        if (keyDeleted) {
-          logger.info(`[update] ${this.name}: stop watching < ${pathKey} >`);
-          this.unsubscribe(splitPath(pathKey));
-        }
+  const watched = React.useRef({});
+  const prevWatched = React.useRef({});
+
+  // On update, update subscriptions
+  React.useEffect(() => {
+    logger.info(`[updating watch] ${name}`);
+
+    Object.keys(watched.current).forEach(pathKey => {
+      const keyExisted = prevWatched.current.hasOwnProperty(pathKey);
+      if (!keyExisted) {
+        logger.info(`[update] ${name}: now watching < ${pathKey} >`);
+        subscribe(splitPath(pathKey));
+      }
+    });
+
+    Object.keys(prevWatched).forEach(pathKey => {
+      const keyDeleted = watched.current.hasOwnProperty(pathKey);
+      if (keyDeleted) {
+        logger.info(`[update] ${name}: stop watching < ${pathKey} >`);
+        unsubscribe(splitPath(pathKey));
+      }
+    });
+
+    prevWatched.current = { ...watched.current };
+    watched.current = {};
+  });
+
+  // On unmount, unsubscribe from everything
+  React.useEffect(() => {
+    return () => {
+      logger.info(`[unmounting]: ${name}`, watched.current);
+
+      Object.keys(prevWatched.current).forEach(pathKey => {
+        logger.info(`[unmount] ${name}: stop watching < ${pathKey} >`);
+        unsubscribe(splitPath(pathKey));
       });
+      logger.debug("store", store);
+      status.current = { unmounted: true };
+    };
+  }, []);
 
-      this.prevWatched = { ...this.watched };
-      this.watched = {};
-    }
-
-    public componentWillUnmount() {
-      logger.info(`[will unmount]: ${this.name}`, this.state);
-      Object.keys(this.state).forEach(pathKey => {
-        logger.info(`[unmount] ${this.name}: stop watching < ${pathKey} >`);
-        this.unsubscribe(splitPath(pathKey));
-      });
+  const watch = valueExtractor(store, watched.current);
 
-      logger.debug("store", this.store);
+  const _subscribe = (path: string[], unsubscribe?: () => void): void => {
+    const pathKey = joinPath(path);
+    watched.current[pathKey] = getValue(path, store.universe);
+    subscribe(path, unsubscribe);
+  };
 
-      this.status.unmounted = true;
-    }
+  const ctx = React.useMemo(() => {
+    const ctx = {
+      dispatch,
+      state,
+      watch,
+      subscribe: _subscribe,
+    };
+    store.plugins.forEach(p => {
+      if (p._internals.viewCtx) {
+        ctx.dispatch = dispatch;
+        p._internals.viewCtx(
+          {
+            ctx,
+            universe: store.universe,
+            comp: {
+              name,
+              compId: compId.current,
+            },
+          },
+          store.config,
+        );
+      }
+    });
+    return ctx;
+  }, [store.universe]);
 
-    public render() {
-      this.createViewCtx();
+  return (func as ((args: any) => (props: any) => any))(ctx)(props);
```

Comment:
I'm pretty sure this breaks support for hooks because `func(ctx)` will be a different function each time.

Reply:
We've done this before and I think it worked. Will do some testing

Reply:
I think hooks still work. At least the update-testing example works as expected.